### PR TITLE
Enhance/5873 - Implement notification about upcoming WordPress version bump

### PR DESCRIPTION
**Pull request by hussain-t**
## Summary

Addresses issue:

- #5873 

## Relevant technical choices

- The IB states the following (it was me who worked on the IB):
>- In `includes/Core/Permissions/Permissions.php`:
>  * Add a new constant `UPDATE_CORE` with the value `googlesitekit_update_core`.
>  * In the `Permissions::__construct` method, update the `UPDATE_CORE` capability with `update_core` to the `base_to_core` array.
>  * Return it in the `Permissions::get_capabilities` method.

> * Check whether the user can update the core using the `hasCapability` selector from the `core/site` datastore and pass the newly added permission constant.

- I initially thought to add the above to the `Permissions` class and use the `hasCapability` selector to check if the user has the `update_core` capability. However, I implemented a better API in the new `Authentication::get_update_core_url()` method that checks whether the user has the `update_core` capability - if so, it returns the URL; otherwise, nothing will be returned. Hence, based on the `updateCoreURL` property, the `Update WordPress` CTA will be displayed.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Core/Authentication/Authentication.php [hunk 1]
```diff
@@ -825,6 +825,27 @@ private function handle_disconnect() {
 		exit();
 	}
 
+	/**
+	 * Gets the update core URL for the user whoc can update the WordPress core.
```
Comment (makiost) [inline includes/Core/Authentication/Authentication.php:829]:
```suggestion
	 * Gets the update core URL if the user can update the WordPress core version.
```

File: assets/js/googlesitekit/datastore/site/info.js [hunk 2]
```diff
@@ -619,6 +627,69 @@ export const selectors = {
...
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state Data store's state.
+	 * @return {(Object|undefined)} WordPress update core URL.
+	 */
+	getUpdateCoreURL: getSiteInfoProperty( 'updateCoreURL' ),
+
+	/**
+	 * Determines whether the current WordPress site's has the minimum required version.
```
Comment (makiost) [inline assets/js/googlesitekit/datastore/site/info.js:652]:
```suggestion
	 * Determines whether the current WordPress site has the minimum required version.
```

File: includes/Core/Authentication/Authentication.php [hunk 3]
```diff
@@ -825,6 +825,27 @@ private function handle_disconnect() {
...
+	 *
+	 * If the multisite is enabled, gets the update core URL for the network admin.
```
Comment (makiost) [inline includes/Core/Authentication/Authentication.php:831]:
```suggestion
	 * If the site is a multisite, it gets the update core URL for the network admin.
```

**Review by makiost — CHANGES_REQUESTED** [review]
Works as expected, just some very minor tweaks to spelling and wording, otherwise it LGTM! 
Great work 👍

File: includes/Core/Authentication/Authentication.php (hunk 1 above)
Comment (hussain-t) [reply includes/Core/Authentication/Authentication.php]: Updated.

File: assets/js/googlesitekit/datastore/site/info.js (hunk 2 above)
Comment (hussain-t) [reply assets/js/googlesitekit/datastore/site/info.js]: Updated.

File: includes/Core/Authentication/Authentication.php (hunk 3 above)
Comment (hussain-t) [reply includes/Core/Authentication/Authentication.php]: Updated, thanks!

**Review by tofumatt — CHANGES_REQUESTED** [review]
There are some issues with the version comparison I spotted, and we only want to render this when the view-only dashboard is being used.

Since this is for the release, can you make sure not to merge `develop` into it? 🙂 I've changed the target branch, but just make sure not to update since this is now targeting `main`.

File: assets/js/components/notifications/WPVersionBumpNotification.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	);
+
+	// The `Update WordPress` CTA should be displayed if the user has `update_core` capability.
+	// The `updateCoreURL` property will be available if the user has the `update_core` capability.
+	// Otherwise, it will be `undefined`. See Authentication::get_update_core_url() method.
+	const updateCoreURL = useSelect( ( select ) =>
+		select( CORE_SITE ).getUpdateCoreURL()
+	);
+
+	if ( hasMinimumWPVersion || version === undefined ) {
```
[the remark this answers is not in the document]
Comment (hussain-t) [reply assets/js/components/notifications/WPVersionBumpNotification.js:70]: I'm rendering the notification component when this is not the view-only dashboard.
https://github.com/google/site-kit-wp/blob/1188ee4d0f464a089f2a3961d08b2bc00f3a9a41/assets/js/components/notifications/BannerNotifications.js#L65-L72

File: assets/js/googlesitekit/datastore/site/info.js [hunk 5]
```diff
@@ -619,6 +627,69 @@ export const selectors = {
...
+	 * Currently, the minimum required version is 5.2.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {string} minVersion The minimum required version.
+	 * @return {(boolean|undefined)} `true` if the WordPress site's version is greater than or equal to the minimum required version, `false` if not. Returns `undefined` if not loaded.
+	 */
+	hasMinimumWordPressVersion: createRegistrySelector(
+		( select ) =>
+			( state, minVersion = '5.2' ) => {
```
[the remark this answers is not in the document]
Comment (hussain-t) [reply assets/js/googlesitekit/datastore/site/info.js:662]: Thanks, let's make it a required param and validate it if that sounds good.

File: assets/js/googlesitekit/datastore/site/info.js [hunk 6]
```diff
@@ -619,6 +627,69 @@ export const selectors = {
...
+					const v2part = parseInt( v2parts[ i ], 10 );
+
+					// Returns `false` if the version is less than the minimum required version.
+					if ( v1part < v2part ) {
+						return false;
+					}
+					// Returns `true` if the version is greater than the minimum required version.
+					if ( v1part > v2part ) {
+						return true;
+					}
```
[the remark this answers is not in the document]
Comment (hussain-t) [reply assets/js/googlesitekit/datastore/site/info.js]: Thanks, @tofumatt. I have updated the selector with a better API.

File: assets/js/components/notifications/WPVersionBumpNotification.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/components/notifications/WPVersionBumpNotification.js:70]: Ohhh, right on, I missed that in the GitHub diff 😓 

Great, thanks! 👍🏻

File: assets/js/googlesitekit/datastore/site/info.test.js [hunk 7]
```diff
@@ -487,5 +489,57 @@ describe( 'core/site site info', () => {
...
+						version: '5.2.0',
+					},
+				} );
+
+				expect(
+					registry
+						.select( CORE_SITE )
+						.hasMinimumWordPressVersion( '5.2' )
+				).toBe( true );
+			} );
```
Comment (tofumatt) [inline assets/js/googlesitekit/datastore/site/info.test.js:542]: Let's add a higher than test here as well.

```suggestion
			it( 'should return `true` if the WordPress version is the same or higher than the provided version', () => {
				provideSiteInfo( registry, {
					wpVersion: {
						major: 5,
						minor: 2,
						version: '5.2.0',
					},
				} );

				expect(
					registry
						.select( CORE_SITE )
						.hasMinimumWordPressVersion( '5.2' )
				).toBe( true );

				expect(
					registry
						.select( CORE_SITE )
						.hasMinimumWordPressVersion( '5.1' )
				).toBe( true );
			} );
```

**Review by tofumatt — APPROVED** [review]
Looks good, just one test to add so it matches the test spec description 🙂

File: assets/js/components/notifications/WPVersionBumpNotification.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	const handleOnCTAClick = useCallback( () => {
+		trackEvent( eventCategory, 'confirm_notification' );
+	}, [ eventCategory ] );
+
+	useMount( () => {
+		trackEvent( eventCategory, 'view_notification' );
+	} );
+
+	const hasMinimumWPVersion = useSelect( ( select ) =>
+		select( CORE_SITE ).hasMinimumWordPressVersion()
```
Comment (tofumatt) [inline assets/js/components/notifications/WPVersionBumpNotification.js:57]: Ah, this needed to be updated but wasn't, so it's failing all of the E2E tests. Here's the fix:

```suggestion
		select( CORE_SITE ).hasMinimumWordPressVersion( '5.2' )
```

**Comment by tofumatt** [comment]
E2E tests keep timing out but some have passed and looks good, so I'm going to merge before heading out for the evening 😅

File: assets/js/components/notifications/WPVersionBumpNotification.js (hunk 8 above)
Comment (hussain-t) [reply assets/js/components/notifications/WPVersionBumpNotification.js]: My bad, I missed it. Thanks for fixing it, @tofumatt 👍

**Pull request merged**